File: server/qe/sample-frameworks/java-selenium-pom/src/main/java/com/nat/base/BasePage.java
package com.nat.base;

import org.openqa.selenium.*;
import org.openqa.selenium.support.ui.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Abstract base class for all Page Objects.
 * Provides common WebDriver interactions with built-in waits,
 * logging, and screenshot capture.
 *
 * <p>Usage: extend this class for every page in your application.
 * <pre>
 * public class LoginPage extends BasePage {
 *     public LoginPage(WebDriver driver) { super(driver); }
 * }
 * </pre>
 */
public abstract class BasePage {

    protected final WebDriver driver;
    protected final WebDriverWait wait;
    private static final Logger log = LoggerFactory.getLogger(BasePage.class);
    private static final int DEFAULT_WAIT_SECONDS = 15;
    private static final String SCREENSHOT_DIR = "target/screenshots/";

    /**
     * Constructs a BasePage with the given WebDriver.
     * Initialises a 15-second explicit wait.
     *
     * @param driver the WebDriver instance for this page
     */
    public BasePage(WebDriver driver) {
        this.driver = driver;
        this.wait   = new WebDriverWait(driver, Duration.ofSeconds(DEFAULT_WAIT_SECONDS));
    }

    /**
     * Navigates the browser to the given URL.
     *
     * @param url the full URL to navigate to
     */
    public void navigateTo(String url) {
        log.info("Navigating to: {}", url);
        driver.get(url);
        waitForPageLoad();
    }

    /**
     * Clicks a web element after waiting for it to be clickable.
     * Scrolls to the element first to ensure it is in the viewport.
     *
     * @param element the element to click
     */
    public void clickElement(WebElement element) {
        wait.until(ExpectedConditions.elementToBeClickable(element));
        scrollToElement(element);
        log.info("Clicking element: <{}>", element.getTagName());
        element.click();
    }

    /**
     * Fills an input field. Clears any existing value before typing.
     *
     * @param element the input element to fill
     * @param value   the value to enter
     */
    public void fillInput(WebElement element, String value) {
        wait.until(ExpectedConditions.visibilityOf(element));
        element.clear();
        element.sendKeys(value);
        log.info("Filled input field '{}' with {} characters",
                element.getAttribute("name") != null ? element.getAttribute("name") : element.getTagName(),
                value.length());
    }

    /**
     * Returns the trimmed visible text of an element,
     * waiting for the element to be visible first.
     *
     * @param element the element to read text from
     * @return trimmed text content
     */
    public String getText(WebElement element) {
        wait.until(ExpectedConditions.visibilityOf(element));
        String text = element.getText().trim();
        log.debug("Got text '{}' from element <{}>", text, element.getTagName());
        return text;
    }

    /**
     * Checks whether an element is currently visible without throwing.
     *
     * @param element the element to check
     * @return true if visible, false otherwise
     */
    public boolean isElementVisible(WebElement element) {
        try {
            return element.isDisplayed();
        } catch (NoSuchElementException | StaleElementReferenceException e) {
            return false;
        }
    }

    /**
     * Waits until document.readyState is "complete".
     * Ensures the page has fully loaded before proceeding.
     */
    public void waitForPageLoad() {
        wait.until(driver -> ((JavascriptExecutor) driver)
                .executeScript("return document.readyState")
                .equals("complete"));
        log.debug("Page load complete: {}", driver.getCurrentUrl());
    }

    /**
     * Scrolls the browser viewport to bring the element into view.
     *
     * @param element the element to scroll to
     */
    public void scrollToElement(WebElement element) {
        ((JavascriptExecutor) driver).executeScript(
                "arguments[0].scrollIntoView({behavior: 'smooth', block: 'center'});",
                element);
    }

    /**
     * Takes a PNG screenshot and saves it to target/screenshots/.
     * Filename includes the given name and a timestamp.
     *
     * @param name a descriptive label for the screenshot file
     */
    public void takeScreenshot(String name) {
        try {
            File screenshotDir = new File(SCREENSHOT_DIR);
            if (!screenshotDir.exists()) {
                screenshotDir.mkdirs();
            }
            String timestamp = LocalDateTime.now()
                    .format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss_SSS"));
            String filename = SCREENSHOT_DIR + name + "_" + timestamp + ".png";
            File srcFile = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
            Files.copy(srcFile.toPath(), Paths.get(filename));
            log.info("Screenshot saved: {}", filename);
        } catch (IOException e) {
            log.error("Failed to save screenshot '{}': {}", name, e.getMessage());
        }
    }

    /**
     * Waits for an element matching the given locator to become visible.
     *
     * @param locator the By locator to wait for
     * @return the visible WebElement
     */
    public WebElement waitForElement(By locator) {
        log.debug("Waiting for element: {}", locator);
        return wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
    }
}

File: server/qe/sample-frameworks/typescript-playwright-pom/src/base/BasePage.ts
import { type Page, type Locator } from '@playwright/test';
import * as fs from 'fs';
import * as path from 'path';

/**
 * Abstract base class for all Page Objects.
 * Provides common Playwright interactions with built-in waits,
 * retry logic, and screenshot capture.
 *
 * @example
 * ```typescript
 * export class LoginPage extends BasePage {
 *   constructor(page: Page) { super(page); }
 * }
 * ```
 */
export abstract class BasePage {
  constructor(protected readonly page: Page) {}

  /**
   * Navigates to the given path (relative to baseURL) and waits
   * for the DOM content to be loaded and network to be idle.
   */
  async navigate(pathOrUrl: string): Promise<void> {
    await this.page.goto(pathOrUrl, { waitUntil: 'domcontentloaded' });
    try {
      await this.page.waitForLoadState('networkidle', { timeout: 3_000 });
    } catch {
      // networkidle is best-effort; domcontentloaded is sufficient
    }
  }

  /**
   * Fills an input field. Clears any existing value with triple-click,
   * then types the new value and verifies it was entered.
   */
  async fillInput(locator: Locator, value: string): Promise<void> {
    await locator.waitFor({ state: 'visible', timeout: 10_000 });
    await locator.click({ clickCount: 3 }); // triple-click selects all
    await locator.fill(value);
    // Verify the value was entered
    const actual = await locator.inputValue();
    if (actual !== value) {
      throw new Error(
        `fillInput: expected '${value}' but input contains '${actual}'`
      );
    }
  }

  /**
   * Clicks an element after waiting for it to be enabled.
   * Scrolls the element into view before clicking.
   */
  async clickElement(locator: Locator): Promise<void> {
    await locator.waitFor({ state: 'visible', timeout: 10_000 });
    await locator.scrollIntoViewIfNeeded();
    await locator.click();
  }

  /**
   * Returns the trimmed innerText of an element,
   * waiting for visibility first.
   */
  async getText(locator: Locator): Promise<string> {
    await locator.waitFor({ state: 'visible', timeout: 10_000 });
    const text = await locator.innerText();
    return text.trim();
  }

  /**
   * Returns true if the element is visible, false otherwise.
   * Never throws — safe to use in conditional logic.
   */
  async isVisible(locator: Locator): Promise<boolean> {
    try {
      return await locator.isVisible();
    } catch {
      return false;
    }
  }

  /**
   * Waits for the page to reach network idle state (up to 10 seconds).
   */
  async waitForNavigation(): Promise<void> {
    await this.page.waitForLoadState('networkidle', { timeout: 10_000 });
  }

  /**
   * Captures a PNG screenshot and saves it to test-results/screenshots/.
   * The filename includes the given name and a timestamp.
   */
  async takeScreenshot(name: string): Promise<void> {
    const dir = path.join(process.cwd(), 'test-results', 'screenshots');
    fs.mkdirSync(dir, { recursive: true });
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const filePath = path.join(dir, `${name}_${timestamp}.png`);
    await this.page.screenshot({ path: filePath, fullPage: false });
    console.log(`Screenshot saved: ${filePath}`);
  }

  /**
   * Scrolls the given locator into the viewport.
   */
  async scrollTo(locator: Locator): Promise<void> {
    await locator.scrollIntoViewIfNeeded();
  }

  /**
   * Hovers over the given element.
   */
  async hover(locator: Locator): Promise<void> {
    await locator.waitFor({ state: 'visible', timeout: 10_000 });
    await locator.hover();
  }

  /**
   * Selects an option from a <select> dropdown by value, label, or index.
   */
  async selectOption(locator: Locator, value: string): Promise<void> {
    await locator.waitFor({ state: 'visible', timeout: 10_000 });
    await locator.selectOption(value);
  }

  /**
   * Sets a file input to the given file path (for file upload inputs).
   */
  async uploadFile(locator: Locator, filePath: string): Promise<void> {
    await locator.waitFor({ state: 'attached', timeout: 10_000 });
    await locator.setInputFiles(filePath);
  }
}
